Standardize tab navigability in web apps forms

nospame opened the pull request:
## Product Description
In line with various web accessibility guidance, only interactive elements will be "focusable" in web apps forms. This means, when navigating with "tab", elements that are selected have some user interaction like an input, button, or link.

## Technical Summary
[USH-4106](https://dimagi.atlassian.net/browse/USH-4106) and notably discussion in its comments. Removes `tabindex` from headings and label text and removes focus for empty `<audio>` and `<video>` elements. Also some small cleanup like removing a `for` attribute from a `<span>` (`for` is only relevant on a `<label>`) and replacing an italic `<i>` with a styled `<span>`.

## Safety Assurance

### Safety story
These changes affect accessibility only and are generally safe _for most users_. Have tested locally with a screen reader and clicking through accessibility tree to see that inputs are correctly labeled.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan
Regression is unlikely, but this should go through a QA cycle for accessibility, to ensure all inputs are properly labeled by their labels and that all interactive elements _are_ focusable.

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change


[USH-4106]: https://dimagi.atlassian.net/browse/USH-4106?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

nospame committed:
Remove invisible audio/video elements from tabindex

nospame committed:
Remove markdown rule giving headings tabindex

Tabindex and keyboard focus should used to signify interactive content

nospame committed:
Remove tabindex from text elements

nospame committed:
Nit style changes in templates.html

nospame committed:
Remove markdown test for heading tabindex

MartinRiese commented on corehq/apps/cloudcare/templates/form_entry/templates.html:
```diff
@@ -266,10 +265,10 @@ <h1 class="title" data-bind="text: title, visible: !showInFormNavigation()"></h1
         <!-- /ko -->
       </div>
-      <i class="hint-text" data-bind="
+      <em class="hint-text" data-bind="
```
Why emphasis on a hint text?

nospame replied:
We use `<i>` pretty widely to mean "icon" and I found it slightly confusing. But you're probably right that "emphasized" isn't correct for simply italic hint text, either. I don't have incredibly strong feelings, does it seem better to leave it as `<i>`?

MartinRiese replied:
I was more wondering if a screen reader uses it somehow. if now. I am fine either way

nospame replied:
It looks like using either `i` or `em` moves it up one level in the accessibility tree, versus using `span` which keeps it at the same level as the label text, so I opted for the latter. https://github.com/dimagi/commcare-hq/pull/34084/commits/45661a5725ad71f95f7ea5da2c1e353a43eb1298

nospame committed:
Style hint text with existing class

nospame committed:
Label map input with aria-labelledby and fix placeholder visibility

div elements like leaflet/mapbox widget aren't automatically labeled by a 'for' attr. And because 'map' only becomes truthy after render, placeholder text was always visible.

nospame committed:
Remove tabindex from non-collapsible group headers

nospame committed:
Improve accessibility for delete repeat button

Adds aria label and makes button clickable with spacebar or enter

nospame committed:
Merge branch 'master' of github.com:dimagi/commcare-hq into em/web-apps-tab-navigability

nospame committed:
Add aria-labelledby to select2 sub-element that gets keyboard focus

By default it labels itself with the placeholder text, which is not what we want

nospame commented:
Requesting review/re-review here please. Thank you!

MartinRiese commented on corehq/apps/cloudcare/static/cloudcare/js/form_entry/entries.js:
```diff
@@ -1343,14 +1344,27 @@ hqDefine("cloudcare/js/form_entry/entries", function () {
+     * Fix that by applying `aria-labelledby` to the element getting keyboard focus.
+     */
+    function applySelect2Labelledby(entryId, isMulti) {
+        var $input = $('#' + entryId);
```
I think it is fair to use `const` and `let` in new code.

nospame replied:
Seems fair to me. https://github.com/dimagi/commcare-hq/pull/34084/commits/3b28124c76f3448b18b59c060bfc3e5561a4593b

nospame committed:
var -> const

nospame committed:
Merge branch 'master' of github.com:dimagi/commcare-hq into em/web-apps-tab-navigability

nospame committed:
Use trans_html_attr

nospame committed:
Merge branch 'master' of github.com:dimagi/commcare-hq into em/web-apps-tab-navigability

MartinRiese commented on corehq/apps/cloudcare/templates/form_entry/templates.html:
```diff
@@ -56,18 +56,12 @@
                       click: deleteRepeat
                   "><i class="fa fa-remove"></i></button>
         </div>
-        <span class="webapp-markdown-output caption" data-bind="html: caption_markdown() || caption(), attr: {id: captionId()}"></span>
```
Is this just duplicate code? Looks very similar to the span and button above.

MartinRiese replied:
thanks for looking into it.
